plasmapy/particles/decorators.py

```diff
@@ -20,488 +18,792 @@
+
+        return particle
 
-        if not particle.is_category(require=require, exclude=exclude, any_of=any_of):
+    parameters_to_skip = ("Z", "mass_numb")
+
+    def perform_pre_validations(self, Z, mass_numb):
+        """
+        Check that there are annotated parameters, that ``Z`` and
+        ``mass_numb`` are integers, and that ``Z`` and ``mass_numb`` are
+        not parameters when more than one parameter is annotated.
+        """
+
+        if not self.parameters_to_process:
             raise ParticleError(
-                _category_errmsg(particle, require, exclude, any_of, funcname)
+                "No parameters have an annotation that will invoke particle_input."
             )
 
-        return particle
+        Z_or_mass_numb = Z is not None or mass_numb is not None
+        multiple_annotated_parameters = len(self.parameters_to_process) > 1
 
-    # The following code allows the decorator to be used either with or
-    # without arguments.  This allows us to invoke the decorator either
-    # as `@particle_input` or as `@particle_input()`, where the latter
-    # call allows the decorator to have keyword arguments.
+        if Z is not None and not isinstance(Z, Integral):
+            raise TypeError("Z must be an integer.")
 
-    if wrapped_function is not None:
-        return decorator(wrapped_function)
-    else:
-        return decorator
+        if mass_numb is not None and not isinstance(mass_numb, Integral):
+            raise TypeError("mass_numb must be an integer.")
+
+        if Z_or_mass_numb and multiple_annotated_parameters:
+            raise ParticleError(
+                "The arguments Z and mass_numb are not allowed when more "
+                "than one argument or keyword is annotated with Particle "
+                "in functions decorated with particle_input."
+            )
+
+    def process_arguments(
+        self, args: tuple, kwargs: Dict[str, Any], instance=None
+    ) -> Dict[str, Any]:
+        """
+        Process the arguments passed to the wrapped callable.
+
+        Parameters
+        ----------
+        args : tuple
+            Positional arguments passed to the wrapped callable.
+
+        kwargs : `dict` of `str` to `object`
+            Keyword arguments provided to the wrapped callable.
+
+        instance : `object`, optional
+            If the wrapped callable is a class instance method, then
+            ``instance`` should be the class instance to which ``func``
+            belongs.
+        """
+
+        arguments = _bind_arguments(self.wrapped, args, kwargs, instance)
+
+        Z = arguments.pop("Z", None)
+        mass_numb = arguments.pop("mass_numb", None)
+
+        self.perform_pre_validations(Z, mass_numb)
+
+        return {
+            parameter: self.process_argument(parameter, argument, Z, mass_numb)
+            for parameter, argument in arguments.items()
+        }
+
+
+def particle_input(
+    wrapped: Optional[Callable] = None,
```

Comment:
```suggestion
    func: Optional[Callable] = None,
```

I'd use `func` here because `func` is the function being wrapped, whereas `wrapped` is the the resulting function after being wrapped.

@namurphy noted that `func` may not be the best choice either since `particle_input` can decorate both functions and methods.  Maybe something the `_callable` is better, with the underscore to avoid namespace conflict with `callable()`